pipeline_server/src/main.rs

```diff
@@ -0,0 +1,873 @@
+        .bind(("127.0.0.1", port))?
+        .run()
+        .await?;
+
+    Ok(())
+}
+
+// `static_files` magic.
+include!(concat!(env!("OUT_DIR"), "/generated.rs"));
+
+fn build_app<T>(app: App<T>, state: WebData<ServerState>) -> App<T>
+where
+    T: ServiceFactory<ServiceRequest, Config = (), Error = ActixError, InitError = ()>,
+{
+    // Creates a dictionary of static files indexed by file name.
+    let generated = generate();
+
+    // Extrac the contents of `index.html`, so we can serve it on the
+    // root endpoint (`/`), while other files are served via the`/static`
+    // endpoint.
+    let index_data = match generated.get("index.html") {
+        None => "<html><head><title>DBSP server</title></head></html>"
+            .as_bytes()
+            .to_owned(),
+        Some(resource) => resource.data.to_owned(),
+    };
+
+    let app = app
+        .app_data(state.clone())
+        .service(list_projects)
+        .service(project_code)
+        .service(project_status)
+        .service(new_project)
+        .service(update_project)
+        .service(compile_project)
+        .service(delete_project)
+        .service(new_config)
+        .service(update_config)
+        .service(delete_config)
+        .service(list_project_configs)
+        .service(new_pipeline)
+        .service(kill_pipeline)
+        .service(delete_pipeline)
+        .service(list_project_pipelines);
+
+    if let Some(static_html) = &state.config.static_html {
+        // Serve static contents from the file system.
+        app.route("/", web::get().to(index))
+            .service(fs::Files::new("/static", static_html).show_files_listing())
+    } else {
+        // Serve static contents embedded in the program.
+        app.route(
+            "/",
+            web::get().to(move || {
+                let index_data = index_data.clone();
+                async { HttpResponse::Ok().body(index_data) }
+            }),
+        )
+        .service(ResourceFiles::new("/static", generated))
+    }
+}
+
+async fn index() -> ActixResult<NamedFile> {
+    Ok(NamedFile::open("static/index.html")?)
+}
+
+fn http_resp_from_error(error: &AnyError) -> HttpResponse {
+    if let Some(db_error) = error.downcast_ref::<DBError>() {
+        let msg = db_error.to_string();
+        match db_error {
+            DBError::UnknownProject(_) => HttpResponse::NotFound(),
+            DBError::OutdatedProjectVersion(_) => HttpResponse::Conflict(),
+            DBError::UnknownConfig(_) => HttpResponse::NotFound(),
+            DBError::UnknownPipeline(_) => HttpResponse::NotFound(),
+        }
+        .body(msg)
+    } else {
+        HttpResponse::InternalServerError().body(error.to_string())
+    }
+}
+
+/// Enumerate the project database.
+///
+/// Returns an array of [project descriptors](`db::ProjectDescr`).
+#[get("/list_projects")]
+async fn list_projects(state: WebData<ServerState>) -> impl Responder {
+    state
+        .db
+        .lock()
```

Comment:
when does unlock happen?

Reply:
when the lock goes out of scope, in this case end of this statement.